sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushDownLeftSemiAntiJoin.scala

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Pushes Left semi and Left Anti joins below the following operators.
+ *  1) Project
+ *  2) Window
+ *  3) Union
+ *  4) Aggregate
+ *  5) Other permissible unary operators. please see [[PushDownPredicate.canPushThrough]].
+ */
+object PushDownLeftSemiAntiJoin extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    // LeftSemi/LeftAnti over Project
+    case Join(p @ Project(pList, gChild), rightOp, LeftSemiOrAnti(joinType), joinCond, hint)
+      if pList.forall(_.deterministic) &&
+        !pList.exists(ScalarSubquery.hasScalarSubquery)&&
```

Comment:
we will remove this line after we finish refactoring the subquery rewrite, right?

Reply:
@cloud-fan We would keep this after the refactoring. Currently , except is planned using anti join. Here is the test that exhibits the problem.
``` sql
SELECT (SELECT min(k) FROM t2 WHERE t2.k = t1.k) min_t2 FROM t1
MINUS
SELECT (SELECT min(k) FROM t2) abs_min_t2 FROM t1 WHERE  t1.k = 'one'
```
After the except operator is replaced .. the plan is : 
```
GlobalLimit 21
+- LocalLimit 21
   +- Project [cast(min_t2#245 as string) AS min_t2#254]
      +- Distinct
         +- Join LeftAnti, (min_t2#245 <=> abs_min_t2#247)
            :- Project [scalar-subquery#244 [(k#242 = k#240)] AS min_t2#245]
            :  :  +- Aggregate [k#242], [min(k#242) AS min(k)#249, k#242]
            :  :     +- Project [k#242]
            :  :        +- LocalRelation [k#242, v#243]
            :  +- Project [k#240, v#241]
            :     +- LocalRelation [k#240, v#241]
            +- Project [scalar-subquery#246 [] AS abs_min_t2#247]
               :  +- Aggregate [min(k#242) AS min(k)#251]
               :     +- Project [k#242]
               :        +- LocalRelation [k#242, v#243]
               +- Filter (k#240 = one)
                  +- Project [k#240, v#241]
                     +- LocalRelation [k#240, v#241]
```
Here we are not pushing down the leftanti operator below project.

Reply:
shouldn't we convert all correlated subqueries to joins before we go to the main optimizer batch?

Reply:
@cloud-fan We will be moving the RewritePredicateSubquery which does the work for converting IN and EXISTS subqueries to Joins. However, Scalar subquery are treated differently in the sense that its handled by a different rule (RewriteCorrelatedScalarSubquery) and are planned using Left outer joins. As part of this work, i wasn't planning on changing Scalar subquery code. Also one thing to note is that, the rule to rewrite scalar subquery is already in the default batch and is run as  a fixedPoint.

Reply:
ah sorry I misread the code. Yea scalar subquery will still be there.

Then why can't we push left anti join through project with scalar subqurey? scalr subquery is similar to a literal which doesn't depend on anything from the child plan.

Reply:
@cloud-fan So letting join pass through scalar subquery ends up with plan like following : 
```
   +- Project [scalar-subquery#244 [(k#242 = k#240)] AS min_t2#245]
            :  +- Aggregate [k#242], [min(k#242) AS min(k)#249, k#242]
            :     +- Project [k#242]
            :        +- LocalRelation [k#242, v#243]
            +- Project [k#240, v#241]
               +- Join LeftAnti, (scalar-subquery#244 [(k#242 = k#240)] <=> abs_min_t2#247)
                  :  +- Aggregate [k#242], [min(k#242) AS min(k)#249, k#242]
                  :     +- Project [k#242]
                  :        +- LocalRelation [k#242, v#243]
                  :- LocalRelation [k#240, v#241]
                  +- Project [scalar-subquery#246 [] AS abs_min_t2#247]
                     :  +- Aggregate [min(k#242) AS min(k)#251]
                     :     +- Project [k#242]
                     :        +- LocalRelation [k#242, v#243]
                     +- Project [k#240, v#241]
                        +- Filter (k#240 = one)
                           +- LocalRelation [k#240, v#241]
```
And things go totally wrong :-). firstly, join ends up hosting a scalar-sub expression which we don't allow (Project, Aggregate, Filter are the only ones that allow). We get a TreeNodeException..

Reply:
@cloud-fan One thing i forgot Wenchen is that, we will be introducing a new rule that pushes down LeftSemi and LeftAnti joins below Join. Once we have that, even though we will skip scalar subqueries in this rule, once they are changed to left outer join after RewriteCorrelatedScalarSubquery, we will be able to push LeftSemi and Left anti joins when applicable. So i think we will get a good plan eventually.

Reply:
`RewriteCorrelatedScalarSubquery` only touches correlated scalar subquery, what about non-correlated scalar subquery? I think it's fine to have non-correlated scalar subquery in the project list, we can still pushdown left anti/semi joins.

Reply:
@cloud-fan Yeah.. We should be able to pushdown in case of non-correlated case. If it is okay with you, i would like to study this and take it as a follow-up. The reason is, at the moment, i don't know if we are generating an optimal plan in this case. Here is the plan before and after for your reference.
Before : 
```
Join LeftSemi, (sum#7L = cast(d#3 as bigint))
:- Project [scalar-subquery#6 [] AS sum#7L]
:  :  +- Aggregate [b#1], [sum(cast(c#2 as bigint)) AS sum#5L]
:  :     +- LocalRelation <empty>, [a#0, b#1, c#2]
:  +- LocalRelation <empty>, [a#0, b#1, c#2]
+- LocalRelation <empty>, [d#3]
```
After the pushdown
```
Project [scalar-subquery#6 [] AS sum#7L]
:  +- Aggregate [b#1], [sum(cast(c#2 as bigint)) AS sum#5L]
:     +- LocalRelation <empty>, [a#0, b#1, c#2]
+- Join LeftSemi, (scalar-subquery#6 [] = cast(d#3 as bigint))
   :  +- Aggregate [b#1], [sum(cast(c#2 as bigint)) AS sum#5L]
   :     +- LocalRelation <empty>, [a#0, b#1, c#2]
   :- LocalRelation <empty>, [a#0, b#1, c#2]
   +- LocalRelation <empty>, [d#3]
```
I wanted to go a little defensive in the first pass. If the plan looks okay to you then
i can make the change. Please let me know.